unittests/AST/ASTImporterTest.cpp

```diff
@@ -2605,6 +2626,112 @@ TEST_P(ImportFunctions,
   EXPECT_EQ(ToM1, ToM2);
 }
 
+class ImportFunctionsVisibilityChain
+    : public ASTImporterTestBase,
+      public ::testing::WithParamInterface<
+          std::tuple<ArgVector, const char *>> {
+protected:
+  ArgVector getExtraArgs() const override { return std::get<0>(GetParam()); }
+  std::string getCode() const { return std::get<1>(GetParam()); }
+};
+
+class ImportFunctionsVisibility
+    : public ASTImporterTestBase,
+      public ::testing::WithParamInterface<
+          std::tuple<ArgVector, std::tuple<const char *, const char *, bool>>> {
+protected:
+  ArgVector getExtraArgs() const override { return std::get<0>(GetParam()); }
+  std::string getCode0() const { return std::get<0>(std::get<1>(GetParam())); }
+  std::string getCode1() const { return std::get<1>(std::get<1>(GetParam())); }
+  bool shouldBeLinked() const { return std::get<2>(std::get<1>(GetParam())); }
+};
+
+const char *ExternF = "void f();";
+const char *StaticF = "static void f();";
+const char *AnonF = "namespace { void f(); }";
+
+TEST_P(ImportFunctionsVisibilityChain, ImportChain) {
+  std::string Code = getCode() + getCode();
+  TranslationUnitDecl *FromTu = getTuDecl(Code, Lang_CXX, "input0.c");
+
+  auto *FromF0 = FirstDeclMatcher<FunctionDecl>().match(
+      FromTu, functionDecl(hasName("f")));
+  auto *FromF1 =
+      LastDeclMatcher<FunctionDecl>().match(FromTu, functionDecl(hasName("f")));
+
+  auto *ToF0 = Import(FromF0, Lang_CXX);
+  auto *ToF1 = Import(FromF1, Lang_CXX);
+
+  EXPECT_TRUE(ToF0);
+  ASSERT_TRUE(ToF1);
+  EXPECT_NE(ToF0, ToF1);
+  EXPECT_EQ(ToF1->getPreviousDecl(), ToF0);
+}
+
+auto ImportChainParams = ::testing::Values(ExternF, StaticF, AnonF);
+
+INSTANTIATE_TEST_CASE_P(ParameterizedTests, ImportFunctionsVisibilityChain,
+                        ::testing::Combine(DefaultTestValuesForRunOptions,
+                                           ImportChainParams), );
+
+TEST_P(ImportFunctionsVisibility, ImportAfter) {
+  TranslationUnitDecl *ToTu = getToTuDecl(getCode0(), Lang_CXX);
+  TranslationUnitDecl *FromTu = getTuDecl(getCode1(), Lang_CXX, "input1.cc");
+
+  auto *ToF0 =
+      FirstDeclMatcher<FunctionDecl>().match(ToTu, functionDecl(hasName("f")));
```

Comment:
Perhaps the `Pattern` definition could go into the test Fixture itself (ImportFunctionsVisibility).